sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/clienttelemetry/ClientMetricsDiagnosticsHandler.java

```diff
@@ -7,29 +7,102 @@
 import com.azure.cosmos.CosmosAsyncClient;
 import com.azure.cosmos.CosmosDiagnosticsContext;
 import com.azure.cosmos.CosmosDiagnosticsHandler;
+import com.azure.cosmos.implementation.Configs;
+import com.azure.cosmos.implementation.ConnectionPolicy;
 import com.azure.cosmos.implementation.ImplementationBridgeHelpers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
 
 import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
 
 public final class ClientMetricsDiagnosticsHandler implements CosmosDiagnosticsHandler {
+    private static final Logger logger = LoggerFactory.getLogger(ClientMetricsDiagnosticsHandler.class);
+    private static final int OPERATION_RECORDING_EXECUTOR_THREAD_MIN_COUNT = 1;
+    private static final int OPERATION_RECORDING_EXECUTOR_THREAD_MAX_COUNT = Configs.getCPUCnt();
+    private static final int OPERATION_RECORDING_EXECUTOR_THREAD_KEEP_ALIVE_TIME_IN_MINUTES = 5;
+
     private static final ImplementationBridgeHelpers.CosmosClientTelemetryConfigHelper.CosmosClientTelemetryConfigAccessor clientTelemetryConfigAccessor =
         ImplementationBridgeHelpers.CosmosClientTelemetryConfigHelper.getCosmosClientTelemetryConfigAccessor();
 
     private final CosmosAsyncClient client;
+    private final ExecutorService executorService;
 
-    public ClientMetricsDiagnosticsHandler(CosmosAsyncClient client) {
+    public ClientMetricsDiagnosticsHandler(CosmosAsyncClient client, ConnectionPolicy connectionPolicy) {
 
         checkNotNull(client, "Argument 'client' must not be null.");
+        checkNotNull(connectionPolicy, "Argument 'connectionPolicy' must not be null.");
+
         this.client = client;
+
+        this.executorService = new ThreadPoolExecutor(
+            OPERATION_RECORDING_EXECUTOR_THREAD_MIN_COUNT,
+            OPERATION_RECORDING_EXECUTOR_THREAD_MAX_COUNT,
+            OPERATION_RECORDING_EXECUTOR_THREAD_KEEP_ALIVE_TIME_IN_MINUTES,
+            TimeUnit.MINUTES,
+            new LinkedBlockingQueue<>(this.getOperationRecordingExecutorQueueSize(connectionPolicy)),
+            new RejectedExecutionHandler() {
+                @Override
+                public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+                    if (!executor.isShutdown()) {
+                        Runnable task = executor.getQueue().poll();
+                        executor.execute(r);
```

Comment:
wouldn't that retry? Why?

Reply:
no, it will not retry -> it basically is the DiscardOldestPolicy + logging, if we ever reach here, it basically means the queue has been full, and usually we would be more interested in newest data